adding warm_start functionality by storing  pop and self._hof if desired

kamalasaurus opened the pull request:
## What does this PR do?

Hi!  So I made a crack at this after v0.6 appeared to push up, here's what I've got so far:

I'm storing/not-rewriting pop and _hof in the case of a warm_start; I -think- that was the general plan.
## Where should the reviewer start?

I've only made additions to tpot/base.py .  The existing nosetests all seem to run fine, but I feel like I should add some others–I'm not 100% sure what to test in terms of the tpot lifecycle and when the stored variables should exist, so I'd appreciate some guidance on that.  Would it be easier to do in the form of a Pull Request?

Also, I looked at driver.py (is this feature also meant to be accessible from the command line?) and thought I'd add something like this:

``` python
    parser.add_argument('--warm-start', action='store', dest='WARM_START', default=False,
        choices=[True, False], type=bool, help='use previously generated models')
```

But I'm not entirely sure if that makes any sense (can the command line pass boolean arguments?  Wouldn't the api be better being something like `-ws` which if added will flag `warm_start` to `True`?) so yeah, some guidance there would also be appreciated as well.  I'm not 100% clear on how tpot is called from the command line either...  does it boot up a new python process whenever you do so?  In which case, is warm_start even possible in that context?

So!  That's where I'm at, hopefully I didn't miss the boat completely!

Let me know what you think at your convenience!
## How should this PR be tested?

Could use some guidance on that!
## What are the relevant issues?
#246
## Questions:
- Do the docs need to be updated?
  Yes, if the warm_start functionality is exposed to the end user.
- Does this PR add new (Python) dependencies?
  Nope.

rhiever commented on tpot/base.py:
```diff
@@ -275,7 +280,8 @@ def fit(self, features, classes):
         self._toolbox.register('evaluate', self._evaluate_individual, features=features, classes=classes)
-        pop = self._toolbox.population(n=self.population_size)
+        # assign population, self._pop can only be not None if warm_start is enabled
+        pop = self._pop if self._pop else self._toolbox.population(n=self.population_size)
```
I know it's less space-efficient, but can you please break this out into a multi-line if/then? Easier to read in the line run.

rhiever commented:
This code looks great! It's exactly what I had in mind for the issue. Thank you for coding it up.

>Also, I looked at driver.py (is this feature also meant to be accessible from the command line?) and thought I'd add something like this:

Warm start functionality unfortunately can't be made available to command-line users since the TPOT instance gets garbage collected as soon as the command-line call ends.

>The existing nosetests all seem to run fine, but I feel like I should add some others–I'm not 100% sure what to test in terms of the tpot lifecycle and when the stored variables should exist, so I'd appreciate some guidance on that.

I think we should make sure to update the unit tests at the very least. There's a test_init unit test that should check to make sure all variables assigned in init are properly assigned to a new TPOT instance (both default and user-provided params).

I think this PR should also add a new unit test that runs a short TPOTClassifier run (1 gen for 20 pop?) on the MNIST practice data set with `warm_start=True` and a fixed `random_state`, checks that the `_pop` and `hof` were assigned to something (e.g. `hof` should have at least one individual in it), runs another small run on that same TPOTClassifier instance (same data set, pop size, gens, different but fixed random state) and does the check again. That test will guarantee, at the very least, that the basic warm start functionality still works.

---

I also think we should consider renaming the hof variable throughout. Can you please change all references to `hof` to `pareto_front` throughout all files? (To make it easier: `hof` shouldn't be referenced in any of the operator files.)

Lastly, please make sure to update the docs (in this PR) with the new warm_start option.

kamalasaurus commented:
Okay, cool~

Thanks for the quick turnaround!  Sorry it took me a couple days to update, I was airplaned for a bit @_@

Anyway!  I think I've addressed all your comments, `hof` is now `pareto_front` in all mentions, docs are updated (I think), tests are added (I think) and the one-liner is now a multi-liner.

The test only has a population of 1, though (is that a facile test?) because when I set it to 20 the unit test became realllllly slow.  It might be good for you to do a basic sanity check (I'm not sure what's necessarily sane in this world!) to make sure everything's fine.  Please look over the test at tests.py:199 as well to make sure my test logic is correct as well.

Let me know what you think!
